src/rowbinary/de.rs

```diff
@@ -1,94 +1,194 @@
 /// `(&[u8], &mut Option<T>) -> Result<usize>`.
 pub(crate) fn deserialize_from<'data, T: Deserialize<'data>>(input: &mut &'data [u8]) -> Result<T> {
-    let mut deserializer = RowBinaryDeserializer { input };
+    // println!("deserialize_from call");
+
+    let mut deserializer = RowBinaryDeserializer {
+        input,
+        validator: (),
+    };
+    T::deserialize(&mut deserializer)
+}
+
+/// Similar to [`deserialize_from`], but expects a slice of [`Column`] objects
+/// parsed from the beginning of `RowBinaryWithNamesAndTypes` data stream.
+/// After the header, the rows format is the same as `RowBinary`.
+pub(crate) fn deserialize_from_and_validate<'data, 'cursor, T: Deserialize<'data>>(
+    input: &mut &'data [u8],
+    columns: &'cursor [Column],
+) -> Result<T> {
+    let mut deserializer = RowBinaryDeserializer {
+        input,
+        validator: DataTypeValidator::new(columns),
+    };
     T::deserialize(&mut deserializer)
 }
 
-/// A deserializer for the RowBinary format.
+/// A deserializer for the RowBinary(WithNamesAndTypes) format.
 ///
 /// See https://clickhouse.com/docs/en/interfaces/formats#rowbinary for details.
-struct RowBinaryDeserializer<'cursor, 'data> {
-    input: &'cursor mut &'data [u8],
+pub(crate) struct RowBinaryDeserializer<'cursor, 'data, Validator = ()>
+where
+    Validator: ValidateDataType,
+{
+    pub(crate) validator: Validator,
+    pub(crate) input: &'cursor mut &'data [u8],
 }
 
-impl<'data> RowBinaryDeserializer<'_, 'data> {
-    fn read_vec(&mut self, size: usize) -> Result<Vec<u8>> {
+impl<'data, Validator> RowBinaryDeserializer<'_, 'data, Validator>
+where
+    Validator: ValidateDataType,
+{
+    pub(crate) fn read_vec(&mut self, size: usize) -> Result<Vec<u8>> {
         Ok(self.read_slice(size)?.to_vec())
     }
 
-    fn read_slice(&mut self, size: usize) -> Result<&'data [u8]> {
+    pub(crate) fn read_slice(&mut self, size: usize) -> Result<&'data [u8]> {
         ensure_size(&mut self.input, size)?;
         let slice = &self.input[..size];
         self.input.advance(size);
         Ok(slice)
     }
 
-    fn read_size(&mut self) -> Result<usize> {
+    pub(crate) fn read_size(&mut self) -> Result<usize> {
         let size = get_unsigned_leb128(&mut self.input)?;
         // TODO: what about another error?
         usize::try_from(size).map_err(|_| Error::NotEnoughData)
     }
 }
 
-#[inline]
-fn ensure_size(buffer: impl Buf, size: usize) -> Result<()> {
-    if buffer.remaining() < size {
-        Err(Error::NotEnoughData)
-    } else {
-        Ok(())
-    }
-}
-
 macro_rules! impl_num {
-    ($ty:ty, $deser_method:ident, $visitor_method:ident, $reader_method:ident) => {
+    ($ty:ty, $deser_method:ident, $visitor_method:ident, $reader_method:ident, $serde_type:expr, $type_hints:expr) => {
         #[inline]
         fn $deser_method<V: Visitor<'data>>(self, visitor: V) -> Result<V::Value> {
+            self.validator.validate($serde_type, $type_hints, 0)?;
```

Comment:
`0` is passed to `validate` from almost every method, cause there is a single case where we need it dynamically - when we are attempting to deserialize FixedString from `deserialize_tuple`, and I could not come up with a more elegant solution for now. Perhaps more `Drop` shenanigans, like it is done for verifying that we deserialized enough Tuple elements?